public/app/features/alerting/unified/hooks/useProduceNewRuleGroup.tsx

```diff
@@ -0,0 +1,102 @@
+import { Action } from '@reduxjs/toolkit';
+import { useCallback } from 'react';
+
+import { dispatch, getState } from 'app/store/store';
+import { RuleGroupIdentifier } from 'app/types/unified-alerting';
+import { PostableRuleDTO, RulerGrafanaRuleDTO, RulerRuleDTO } from 'app/types/unified-alerting-dto';
+
+import { alertRuleApi } from '../api/alertRuleApi';
+import { addRuleAction, deleteRuleAction, pauseRuleAction, ruleGroupReducer } from '../reducers/ruler/ruleGroups';
+import {
+  fetchPromAndRulerRulesAction,
+  fetchRulesSourceBuildInfoAction,
+  getDataSourceRulerConfig,
+} from '../state/actions';
+
+export function useProduceNewRuleGroup() {
+  const [fetchRuleGroup, _fetchRuleGroupState] = alertRuleApi.endpoints.getRuleGroupForNamespace.useLazyQuery();
+  const [deleteRuleGroup, _deleteRuleGroupState] = alertRuleApi.endpoints.deleteRuleGroupFromNamespace.useLazyQuery();
+  const [updateRuleGroup, updateRuleGroupState] = alertRuleApi.endpoints.updateRuleGroupForNamespace.useMutation();
+
+  const produceNewRuleGroup = async (ruleGroup: RuleGroupIdentifier, action: Action) => {
+    const { ruleSourceName, group, namespace } = ruleGroup;
+
+    // @TODO we should really not work with the redux state (getState) here
+    await dispatch(fetchRulesSourceBuildInfoAction({ rulesSourceName: ruleSourceName }));
+    const rulerConfig = getDataSourceRulerConfig(getState, ruleSourceName);
+
+    const currentRuleGroup = await fetchRuleGroup({ rulerConfig, namespace, group }).unwrap();
+
+    // @TODO convert rule group to postable rule group – TypeScript is not complaining here because
+    // the interfaces are compatible but it _should_ complain
+    const newRuleGroup = ruleGroupReducer(currentRuleGroup, action);
+
+    // if this was the last rule in the group after applying the action on the reducer we remove the group in its entirety
+    const deleteEntireGroup = newRuleGroup.rules.length === 0;
+    if (deleteEntireGroup) {
+      return deleteRuleGroup({
+        rulerConfig,
+        namespace,
+        group,
+      });
+    }
+
+    // otherwise assume we just want to update the existing group
+    return updateRuleGroup({
+      rulerConfig,
+      namespace,
+      payload: newRuleGroup,
+    }).unwrap();
+  };
+
+  // @TODO merge loading state with the fetching state
+  return [produceNewRuleGroup, updateRuleGroupState] as const;
+}
+
+export function usePauseRuleInGroup() {
+  const [produceNewRuleGroup, updateState] = useProduceNewRuleGroup();
+
+  const pauseFn = useCallback(
+    async (ruleGroup: RuleGroupIdentifier, rule: RulerGrafanaRuleDTO, pause: boolean) => {
+      const uid = rule.grafana_alert.uid;
+      const action = pauseRuleAction({ uid, pause });
+
+      await produceNewRuleGroup(ruleGroup, action);
+    },
+    [produceNewRuleGroup]
+  );
```

Comment:
(At first glance) it seems like there's more flexibility in here than you might want. Within the spirit of the changes, perhaps we want to abstract it away even more so that there's no way you could forget to use `produceNewRuleGroup` - but that might not solve that problem either

Reply:
Good point – I don't know of a good way to make it mandatory (without a human factor involved) to call this hook but hopefully the implementation of these hooks is so small that it is easy to manually review these and catch hooks not using `useProduceNewRuleGroup`